Unity-Technologies/Graphics
# New Shader Graph serialization format

### Purpose of this PR
The way that polymorphic serialization is currently handled in Shader Graph greatly increases the likelihood of merge conflicts, and makes it much harder to resolve merge conflicts.

The short of how it works currently is that anything that needs polymorphism such as the list of nodes are stored as nested JSON. Since JSON doesn’t allow actual line-breaks in strings – they get converted to `\n` – all data for a single node is stored in 1 line. To make matters worse, each node also stores each of its ports like this. It’s very hard to resolve merge conflicts in these strings due to the fact that it’s all in one line, and that all quotes are escaped. Furthermore, there’s no chance of JSON support from any text editor, because it’s inside a string.

This PR introduces a new file format that can store multiple JSON objects, and allows them to reference each other. A JSON object is a class that inherits a base class `JsonObject`, which provides an ID (`objectId`) stored as a string. This ID is read-only, and auto-generated on construction or filled in during deserialization. It consists only of letters and numbers, and can therefore be used fairly easily during shader code gen.

Serialized fields containing references to other JSON objects are required to use a struct `JsonData<T>` or `JsonRef<T>`. `T` is required to be a `JsonObject`. This struct has implicit conversions from/to type T. It contains a non-serialized field for storing a regular C# reference to the `T` value, and a serialized string field containing the ID of that value. The difference between `JsonData<T>` and `JsonRef<T>` is that the former implies ownership of the value, while the latter merely references it.

Deserialization is performed in two stages. First, we scan the file for JSON objects, gathering their IDs and types. The first object of the file is assumed to be the main object. This is used to construct each object, and create a mapping from ID to object. We then loop through each object, and use `EditorJsonUtility.FromJsonOverwrite` to fill them with values from the file. `JsonRef<T>` and `JsonData<T>` uses `ISerializationCallbackReceiver` to hook into the deserialization process. It accesses the ID to object map from before, and uses that to turn the serialized ID into an actual C# reference.

For serialization, we start from a main object (which would typically be `GraphData`, but could also be e.g. `CopyPasteGraph`), and use `EditorJsonUtility.ToJson` to serialize it. `JsonData<T>` also takes part in the serialization process, and when serialized will access a global context, and make sure that the referenced value goes into the file.

`JsonObject`s provide a few virtual methods for hooking into the (de)serialization process. These are fairly simple for now, and could be extended for future needs:

```csharp
void OnBeforeSerialize()
```
```csharp
void OnAfterDeserialize()
```
Works like the usual `ISerializationCallbackReceiver` callbacks.

```csharp
void OnAfterDeserialize(string json)
```
Like `OnAfterDeserialize()`, but allows you to read the JSON string for the object being deserialized for upgrade purposes. Object references can be assumed to be stable, but they might not have been deserialized yet, so don't try to use the values in them.

```csharp
void OnAfterMultiDeserialize(string json)
```
Like `OnAfterDeserialize(string json)`, but after all objects have been deserialized. Note that you can still run into order issues if you're trying to access something that requires some setup to run in another `OnAfterMultiDeserialize(string)`.

---
### Testing status

**Manual Tests**
- Imported various test projects from scratch
- Opened existing shader graphs to make sure that they still work
- Re-saved existing graphs to new format, and made sure that they still work
- Created and opened new shader graphs

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)
Added a folder outside the Assets folder which contains a number of 7.x and 8.x saved graphs. Added a test to copy contents over to the Assets folder, import the created asset, deserialize the graph and run Validate on the graph.

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Fhappy-json-v2

---
### Comments to reviewers
Notes for the reviewers you have assigned.
